test/boost/cql_query_test.cc

```diff
@@ -361,6 +361,51 @@ SEASTAR_TEST_CASE(test_list_of_tuples_with_bound_var) {
     });
 }
 
+SEASTAR_TEST_CASE(test_bound_var_in_collection_literal) {
+    return do_with_cql_env_thread([](cql_test_env& e) {
+        e.execute_cql("create table list_t (pk int PRIMARY KEY, c1 list<int>);").get();
+        e.execute_cql("create table set_t (pk int PRIMARY KEY, c1 set<int>);").get();
+        e.execute_cql("create table map_t (pk int PRIMARY KEY, c1 map<int, int>);").get();
+
+        auto insert_list = e.prepare("insert into list_t (pk, c1) values (112, [997, ?])").get0();
+        auto insert_set = e.prepare("insert into set_t (pk, c1) values (112, {997, ?})").get0();
+        auto insert_map_key = e.prepare("insert into map_t (pk, c1) values (112, {997: 112, ?: 112})").get0();
+        auto insert_map_value = e.prepare("insert into map_t (pk, c1) values (112, {997: 112, 112: ?})").get0();
+        BOOST_REQUIRE_THROW(
+            e.prepare("insert into map_t (pk, c1) values (112, {997: 112, ?})").get0(),
+            exceptions::syntax_exception
+        );
+
+        std::vector<cql3::prepared_cache_key_type> stmts = {insert_list, insert_set, insert_map_key, insert_map_value};
+
+        for (const auto& stmt : stmts) {
+            const auto null_value = cql3::raw_value::make_null();
+            BOOST_REQUIRE_THROW(
+                e.execute_prepared(stmt, {null_value}).get(),
+                exceptions::invalid_request_exception
+            );
+
+            const auto short_value = cql3::raw_value::make_value(short_type->decompose(data_value((int16_t)69)));
+            BOOST_REQUIRE_THROW(
+                e.execute_prepared(stmt, {short_value}).get(),
+                exceptions::invalid_request_exception
+            );
+
+            const auto unset_value = cql3::raw_value::make_unset_value();
+            BOOST_REQUIRE_THROW(
+                e.execute_prepared(stmt, {unset_value}).get(),
+                exceptions::invalid_request_exception
+            );
+
+            const auto int_value = cql3::raw_value::make_value(short_type->decompose(data_value(69)));
+            BOOST_REQUIRE_THROW(
+                e.execute_prepared(stmt, {int_value}).get(),
+                exceptions::invalid_request_exception
+            );
+        }
```

Comment:
What about testing a **successful** case?
If the goal of this test is only testing errors, maybe you should clarify this in its name and/or a comment.